test/python/circuit/library/test_permutation.py

```diff
@@ -39,5 +42,125 @@ def test_permutation_bad(self):
+                [0, 0, 0, 0, 0, 0, 1, 0],
+                [0, 1, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 1, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 1, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 1],
+            ]
+        )
+        self.assertTrue(np.array_equal(perm.__array__(dtype=int), expected_op))
+
+    def test_pattern(self):
+        """Test the ``pattern`` method."""
+        pattern = [1, 3, 5, 0, 4, 2]
+        perm = PermutationGate(pattern)
+        self.assertTrue(np.array_equal(perm.pattern, pattern))
+
+    def test_inverse(self):
+        """Test correctness of the ``inverse`` method."""
+        perm = PermutationGate([1, 3, 5, 0, 4, 2])
+
+        # We have the permutation 1->0, 3->1, 5->2, 0->3, 4->4, 2->5.
+        # The inverse permutations is 0->1, 1->3, 2->5, 3->0, 4->4, 5->2, or
+        # after reordering 3->0, 0->1, 5->2, 1->3, 4->4, 2->5.
+        inverse_perm = perm.inverse()
+        expected_inverse_perm = PermutationGate([3, 0, 5, 1, 4, 2])
+        self.assertTrue(np.array_equal(inverse_perm.pattern, expected_inverse_perm.pattern))
+
+
+class TestPermutationGatesOnCircuit(QiskitTestCase):
+    """Tests for quantum circuits containing permutations."""
+
+    def test_append_to_circuit(self):
+        """Test method for adding Permutations to quantum circuit."""
+        qc = QuantumCircuit(5)
+        qc.append(PermutationGate([1, 2, 0]), [0, 1, 2])
+        qc.permutation([2, 3, 0, 1], [1, 2, 3, 4])
+        self.assertIsInstance(qc.data[0].operation, PermutationGate)
+        self.assertIsInstance(qc.data[1].operation, PermutationGate)
+
+    def test_inverse(self):
+        """Test inverse method for circuits with permutations."""
+        qc = QuantumCircuit(5)
+        qc.permutation([1, 2, 3, 0], [0, 4, 2, 1])
+        qci = qc.inverse()
+        qci_pattern = qci.data[0].operation.pattern
+        expected_pattern = [3, 0, 1, 2]
+
+        # The inverse permutations should be defined over the same qubits but with the
+        # inverse permutation pattern.
+        self.assertTrue(np.array_equal(qci_pattern, expected_pattern))
+        self.assertTrue(np.array_equal(qc.data[0].qubits, qci.data[0].qubits))
+
+    def test_reverse_ops(self):
+        """Test reverse_ops method for circuits with permutations."""
+        qc = QuantumCircuit(5)
+        qc.permutation([1, 2, 3, 0], [0, 4, 2, 1])
+        qcr = qc.reverse_ops()
+
+        # The reversed circuit should have the permutation gate with the same pattern and over the
+        # same qubits.
+        self.assertTrue(np.array_equal(qc.data[0].operation.pattern, qcr.data[0].operation.pattern))
+        self.assertTrue(np.array_equal(qc.data[0].qubits, qcr.data[0].qubits))
+
+    def test_conditional(self):
+        """Test adding conditional permutations."""
+        qc = QuantumCircuit(5, 1)
+        qc.permutation([1, 2, 0], [2, 3, 4]).c_if(0, 1)
+        self.assertIsNotNone(qc.data[0].operation.condition)
+
+    def test_qasm(self):
```

Comment:
Can you add a similar test for qpy serialization. Looking at the gate definition I don't expect any issues since the permutation pattern is defined in the gate parameters so qpy should just pass the list of ints to the constructor and everything should work fine, but it'd be good to validate it is recreating it exactly. (it'd  be fine to add here or in `test/python/qpy`)

Reply:
Done in [e577372](https://github.com/Qiskit/qiskit-terra/pull/9157/commits/e57737220bb6adeebc7c8fcd111943b88f3920ce)